Added corporate proxy support and option for maintaining specific target path

uxkjaer opened the pull request.

uxkjaer committed:
Update install.js

Added StreamZip to handle extraction and tunnel for proxy agent

uxkjaer committed:
Update README.md

uxkjaer committed:
Update package.json

uxkjaer committed:
ESLINT fixes

bastilimbach commented on README.md:
```diff
@@ -10,6 +10,14 @@ $ yarn add -D sapui5-runtime
 
+### Proxy
+If you are using a proxy, the proxy needs to be set as an environment variable. Setting in the node config isn't enough. 
+[Linux / MACOS]
```
```suggestion
[Linux / macOS]
```

bastilimbach commented on README.md:
````diff
@@ -10,6 +10,14 @@ $ yarn add -D sapui5-runtime
+### Proxy
+If you are using a proxy, the proxy needs to be set as an environment variable. Setting in the node config isn't enough. 
+[Linux / MACOS]
+``` export https_proxy http(s)://username:password@hostname:port```
````
Axios should be able to deal with the https proxy automatically as it looks for the `https_proxy` environment variable. Tunnel is only used/useful if you only want to proxy https resources over http. To achieve this you used the `httpsOverHttp` function. In case of https over https tunnel isn't used.

I would purpose the following:
```suggestion
``` export http_proxy http://username:password@hostname:port```
``` export http_proxys https://username:password@hostname:port```
```

Then we check in `install.js` line 11 if http_proxy is set and https_proxy is not set. In this case we want to use Tunnel to proxy httpsOverHttp. What do you think?

bastilimbach commented on package.json:
```diff
@@ -28,7 +28,8 @@
   },
   "dependencies": {
-    "adm-zip": "^0.4.7",
+    "node-stream-zip": "^1.8.2",
```
Is there a specific reason why you changed the dependency to deal with zip archives?

uxkjaer replied:
Hi Sebastian, I had some issues after downloading the zip archives, it gave me an unexpected end of zip file error. When I changed to node-stream-zip, i had no issues ad all with my testing

bastilimbach replied:
Mh, this error usually occurs if the downloaded ZIP file was corrupted. Did you try it multiple times? Maybe the download didn't work. If this happens every time there might be a problem with windows. I'm not able to test it on windows. Therefore I would be fine with changing the module, if this error is reproducible.

uxkjaer replied:
Yes i did try multiple times and it errorred, I had no issues with my testing using node-stream

bastilimbach replied:
Alright, then I’m fine with using node-stream instead of adm zip 👍

bastilimbach commented on src/install.js:
```diff
@@ -1,10 +1,29 @@
+const tunnel = require('tunnel')
+
+let agent
+if (process.env.HTTPS_PROXY) {
```
```suggestion
if (!process.env.HTTPS_PROXY && process.env.HTTP_PROXY ) {
```

See comment in `README.md`

bastilimbach commented on src/install.js:
```diff
@@ -1,10 +1,29 @@
+  const urlArray = urlString.split('@')
+  const proxyAuth = urlArray[0]
+  const hostname = urlArray[1].split(':')[0]
+  const port = urlArray[1].split(':')[1]
```
I would prefer to initialize a new URL: https://nodejs.org/api/url.html#url_constructor_new_url_input_base

With this, we can more safely and easily access the separate URL parts. E.g:
```javascript
const myURL = new URL('https://example.org/foo#bar');
console.log(myURL.hash);
// Prints #bar

const myURL = new URL('https://example.org:81/foo');
console.log(myURL.host);
// Prints example.org:81
```

uxkjaer replied:
I've added the new logic. Please review.

bastilimbach commented on src/install.js:
```diff
@@ -54,13 +75,26 @@ async function downloadSAPUI5(downloadURL, downloadPath) {
+          Cookie: 'eula_3_1_agreed=tools.hana.ondemand.com/developer-license-3_1.txt',
+        },
+      })
+    }
```
As the axios config is quite similar I would prefer something like:
```javascript
const requestConfig = {
  responseType: 'stream',
  headers: {
    Cookie: 'eula_3_1_agreed=tools.hana.ondemand.com/developer-license-3_1.txt',
  },
}

if (agent) {
  requestConfig.httpsAgent = agent
  requestConfig.proxy = false
}

const response = await axios.get(downloadURL, requestConfig)
```

uxkjaer replied:
Agree, implemented

bastilimbach commented on src/install.js:
```diff
@@ -75,21 +109,47 @@ async function downloadSAPUI5(downloadURL, downloadPath) {
+      })
+      response.data.on('error', () => {
+        reject()
+      })
```
Can be shortened to:
```javascript
response.data.on('error', reject)
```

bastilimbach commented on src/install.js:
```diff
@@ -75,21 +109,47 @@ async function downloadSAPUI5(downloadURL, downloadPath) {
+    zip.extract(null, targetPath, () => {
+      zip.close()
+    })
+  })
```
As mentioned in `package.json` I'm not sure why you changed the module from AdmZip to StreamZip. Does it have a performance improvement?

uxkjaer replied:
Please see my comments above

bastilimbach commented on src/install.js:
```diff
@@ -99,14 +159,17 @@ async function installSAPUI5() {
+      extractArchive(sapui5Archive, libDir)
+    }).catch((error) => {
+      console.log(error.message)
+    })
```
Why did you introduce a separate `then` and `catch` here? If `downloadSAPUI5` throws an error or rejects the promise, the catch block of the `try, catch` will be executed and print the error.
What is the reason for this change? I might be wrong here as I didn't test it locally. Maybe the catch is required.

uxkjaer replied:
It was due to my testing of the unexpected archive. It's reverted back

bastilimbach commented on src/install.js:
```diff
@@ -75,21 +109,47 @@ async function downloadSAPUI5(downloadURL, downloadPath) {
 }
 
-const libDir = path.resolve(`${__dirname}/../lib`)
+const libDir = (config.dest) ? path.resolve(`${__dirname}/../../../`, config.dest) : path.resolve(`${__dirname}/../lib`)
```
What would be the use-case here? I cannot come up with a use-case where such a configuration would be useful. Maybe you have some specific case which you can share for me to better understand why we would need such a configuration.

Nevertheless, if there is no real use-case for such a scenario I would prefer to not implement this config for the following reasons:

1. If the user sets the `dest` config he/she cannot require the module anymore with `const sapui5 = require('sapui5-runtime')` as we return a hardcoded path to the lib folder. Take a look at [`index.js`](https://github.com/bastilimbach/sapui5-runtime/blob/master/src/index.js). If we want to support the config, we also need to adapt this file to return the correct path.
2. We might run into file authorization errors if the user provides a path to a restricted folder on his/her system
3. We don't check which path is entered so it is possible to mistakingly add relative paths like: `../../../../` which would lead to a security vulnerability. Also, the libDir is always deleted from the system before running the installation which would corrupt the file system of the end-user if such a relative path is entered.

uxkjaer replied:
I understand your concerns. My purpose is to download the sapui5 runtime once and then have in a shared spot for all apps. I guess you could install the package globally instead maybe?
My usecase is to develop locally for all our apps 20+ and use one runtime instead of having to download it for every project.

bastilimbach replied:
Ah ok, I get it.
This package is intended to be used as a local package and not a global one. For such a use-case I would either download the file manually and place it somewhere safe or write a plugin for the [UI5 tooling](https://github.com/SAP/ui5-tooling) which can be executed globally. This package should be used on a project by project basis as multiple UI5 projects typically require different SAPUI5 versions. What do you think?

uxkjaer replied:
Yeah I understand your comment. I guess you could install it globally instead and just use the lib. I'll revert those changes.

bastilimbach reviewed (changes requested):
Thank you so much for the Pull Request! I really appreciate the time you took to implement the missing proxy feature. I hope you can take a look at my review and incorporate the requested changes. As soon as we merge this PR I'll publish a new version to NPM 👍

Btw. please remove the `package-lock.json` and add/remove dependencies using [yarn](https://yarnpkg.com/en/).

Again, thank you very much for your effort! 🎉

uxkjaer committed:
Update README.md

Co-Authored-By: Sebastian Limbach <bastilimbach@users.noreply.github.com>

uxkjaer committed:
Update README.md

Co-Authored-By: Sebastian Limbach <bastilimbach@users.noreply.github.com>

uxkjaer committed:
Update src/install.js

Co-Authored-By: Sebastian Limbach <bastilimbach@users.noreply.github.com>

uxkjaer committed:
Update install.js

Used Url syntax

uxkjaer committed:
Update install.js

uxkjaer committed:
Update install.js

uxkjaer committed:
Update install.js

uxkjaer committed:
Added changes and tested with proxy

uxkjaer committed:
Updated to latest eslint to get rid of URL no-undef error

bastilimbach commented on src/install.js:
```diff
@@ -1,10 +1,29 @@
+  }
+  agent = tunnel.httpsOverHttp({
+    proxy,
+  })
```
I would prefer:
```javascript
agent = tunnel.httpsOverHttp({
  proxy: {
    host: httpsProxyObject.hostname,
    port: httpsProxyObject.port,
    proxyAuth: httpsProxyObject.auth
  },
})
```

Take a look at: https://nodejs.org/api/url.html#url_url_strings_and_url_objects
Also tunnel checks if proxyAuth is undefined: https://github.com/koichik/node-tunnel/blob/a98da54560203b05424e44f9cc9d759df0aa3197/lib/tunnel.js#L119

bastilimbach commented on src/install.js:
```diff
@@ -75,21 +104,45 @@ async function downloadSAPUI5(downloadURL, downloadPath) {
+  zip.on('ready', () => {
+    console.log(`All entries read: ${zip.entriesCount}`)
+
+    const progressBar1 = new ProgressBar('Extracting: [:bar]', {
```
I would prefer to call it something like `extractionProgressBar`.

uxkjaer committed:
Update install.js

Added agent = tunnel.httpsOverHttp({
  proxy: {
    host: httpsProxyObject.hostname,
    port: httpsProxyObject.port,
    proxyAuth: httpsProxyObject.auth
  },
})

uxkjaer committed:
Update install.js

uxkjaer committed:
Update install.js

uxkjaer committed:
Update README.md

bastilimbach commented:
I tested it locally and it works really great 👍 Could you please fix the last eslint errors? Then I will merge the PR and publish a new npm version of this package

bastilimbach reviewed (approved):
I tested it locally and it works really great 👍 Could you please fix the last eslint errors? Then I will merge the PR and publish a new npm version of this package.

Btw you can fix the error automatically by running `npm run lint:fix`

bastilimbach commented:
I found some free time and fixed the last errors myself. Thanks again for adding the proxy support! 🎉
